s3/src/main/scala/tamer/s3/S3Configuration.scala

```diff
@@ -0,0 +1,119 @@
+package tamer
+package s3
+
+import com.sksamuel.avro4s.{Codec, SchemaFor}
+import eu.timepit.refined.types.numeric.PosInt
+import zio.stream.ZTransducer
+import zio.{Queue, UIO, ZIO}
+
+import java.time.format.DateTimeFormatter
+import java.time.{Duration, Instant}
+import scala.util.hashing.MurmurHash3.stringHash
+
+
+final case class S3Configuration[
+  K <: Product : Codec : SchemaFor,
+  V <: Product : Codec : SchemaFor,
+  S <: Product : Codec : SchemaFor
+](
+   bucketName: String,
+   prefix: String,
+   tamerStateKafkaRecordKey: Int,
+   transducer: ZTransducer[Any, TamerError, Byte, V],
+   parallelism: PosInt,
+   timeouts: S3Configuration.Timeouts,
+   transitions: S3Configuration.StateTransitions[K, V, S],
+ ) {
+  val generic: SourceConfiguration[K, V, S] = SourceConfiguration[K, V, S](
+    SourceConfiguration.SourceSerde[K, V, S](),
+    defaultState = transitions.initialState,
+    tamerStateKafkaRecordKey = tamerStateKafkaRecordKey,
+    S3Configuration.this.toString,
+  )
+}
+
+object S3Configuration {
+
+  case class Timeouts(
+                       minimumIntervalForBucketFetch: Duration,
+                       maximumIntervalForBucketFetch: Duration,
+                     )
+
+  case class StateTransitions[K, V, S](
+                                        initialState: S,
+                                        getNextState: (KeysR, S, Queue[Unit]) => UIO[S],
+                                        deriveKafkaRecordKey: (S, V) => K,
+                                        selectObjectForState: (S, Keys) => Option[String]
+                                      )
```

Comment:
`StateTransitions` is a bit imprecise. For example `deriveKafkaRecordKey` might use state as input information to create a key. It might also ignore it completely. But at the end of the day it does not contribute to the state transition. Similarly `selectObjectForState`, while it defines the behavior of a particular state, it can't influence state transition.